### PR TITLE
Don't use `this` in reporter

### PR DESCRIPTION
**Pull request by SimenB**
Missing tests and grouping by file.

The responsibility of that should be in the reporter, for now I just deleted it from `done`

### CODE REVIEW

**Comment by SimenB** [comment]
@wojciechczerniak This is ready for review now 😄

File: src/core/parse.js [hunk 1]
```diff
@@ -14,7 +14,7 @@ var lineEndingsRe = /\r\n|\n|\r/gm
 var parse = function( err, res ) {
 	if ( err ) { throw new Error( err ) }
 
-	return res.forEach( function( file, i ) {
+	res.forEach( function( file, i ) {
```
Comment (SimenB) [inline src/core/parse.js]: `forEach` is void, nothing happens on return

File: src/core/parse.js [hunk 2]
```diff
@@ -41,7 +41,7 @@ var parse = function( err, res ) {
 
 		// if on the last file, call the done function to output success or error msg
 		if ( this.cache.fileNo === res.length - 1 ) {
-			return this.reporter( '', 'done' )
+			this.done()
```
Comment (SimenB) [inline src/core/parse.js]: `done` calls the reporter

File: src/core/reporter.js [hunk 3]
```diff
@@ -2,72 +2,87 @@
...
-		this.cache.msg = 'Stylint: ' + numOfErrors + ' Errors.'
-		this.cache.msg += this.config.maxErrors ? ' (Max Errors: ' + this.config.maxErrors + ')' : ''
-
-		this.cache.msg += '\nStylint: ' + numOfWarnings + ' Warnings.'
-		this.cache.msg += this.config.maxWarnings ? ' (Max Warnings: ' + this.config.maxWarnings + ')' : ''
-
-		// if you set a max it kills the linter
-		if ( kill === 'kill' ) {
-			this.cache.msg += '\nStylint: Over Error or Warning Limit.'
+// TODO: Structure input to this method better (#366)
```
Comment (SimenB) [inline src/core/reporter.js]: I'll do this now

File: test/core/reporter.test.js [hunk 4]
```diff
@@ -0,0 +1,83 @@
+'use strict'
+
+var assert = require( 'assert' )
+var stripColor = require( 'chalk' ).stripColor
+var reporter = require( '../../src/core/reporter' )
+
+function genMessage( file, rule, severity ) {
```
Comment (SimenB) [inline test/core/reporter.test.js]: might pull these out into a util of some kind later

File: src/core/done.js [hunk 5]
```diff
@@ -1,90 +1,41 @@
...
-				var file = output.file
-
-				if ( groupedByFile.hasOwnProperty( file ) ) {
-					groupedByFile[file].push( output )
-				}
-				else {
-					groupedByFile[file] = [output]
-				}
-			} )
+	var message = this.reporter( this.cache.messages, shouldKill, {
```
Comment (wojciechczerniak) [inline src/core/done.js]: Why some options are inside `options` object and some on their own as arguments?

File: src/core/reporter.js [hunk 6]
```diff
@@ -1,73 +1,82 @@
...
-		var violations = groupBy( this.cache.violations, 'severity' )
-		var numOfErrors = violations.Error ? violations.Error.length : 0
-		var numOfWarnings = violations.Warning ? violations.Warning.length : 0
-
-		this.cache.msg = 'Stylint: ' + numOfErrors + ' Errors.'
-		this.cache.msg += this.config.maxErrors ? ' (Max Errors: ' + this.config.maxErrors + ')' : ''
-
-		this.cache.msg += '\nStylint: ' + numOfWarnings + ' Warnings.'
-		this.cache.msg += this.config.maxWarnings ? ' (Max Warnings: ' + this.config.maxWarnings + ')' : ''
+// TODO: Structure input to this method better (#366)
```
Comment (wojciechczerniak) [inline src/core/reporter.js]: Oh, I see why. Nevermind then 😄

**Pull request merged**